x/pairing/keeper/scores/score.go

```diff
@@ -0,0 +1,222 @@
+//
+// A pairing score describes the suitability of a provider for a pairing slot (under a given strategy).
+// The score depends on the slot's requirements: for example, given a slot which requires geolocation in
+// Asia, a provider in Asia will generally get higher score than one in Europe. The score is calculated for
+// each <provider, slot> combination.
+//
+// A pairing score strategy defines the weight of each score requirement in the final score calculation
+// for a <provider, slot> combination. For example, given a slot with several requirements, then the
+// overall pairing score would be calculated as score1^w1 + score2^w2 + ... (where score1 is the score
+// of the provider with respect to the first requirement, score2 with respect to the second requirement
+// and so on).
+//
+//
+// To add a new requirement, one should create a new object that satisfies the ScoreReq interface and update the
+// CalcSlots() function so the new requirement will be assigned to the pairing slots (according to some logic).
+// Lastly, append the new requirement object to the allReqTypes var in the init() function in this file. Use
+// StakeReq or GeoReq as examples.
+
+package scores
+
+import (
+	"fmt"
+	"math/big"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	commontypes "github.com/lavanet/lava/common/types"
+	epochstoragetypes "github.com/lavanet/lava/x/epochstorage/types"
+	planstypes "github.com/lavanet/lava/x/plans/types"
+	tendermintcrypto "github.com/tendermint/tendermint/crypto"
+)
+
+var (
+	uniformStrategy ScoreStrategy
+	allReqNames     []string
+)
+
+// TODO: currently we'll use weight=1 for all reqs. In the future, we'll get it from policy
+func init() {
+	// gather all req names to a list
+	allReqNames = []string{stakeReqName}
+
+	// init strategy
+	uniformStrategy = make(ScoreStrategy)
+	for _, reqName := range allReqNames {
+		uniformStrategy[reqName] = 1
+	}
+
+	if len(allReqNames) != len(uniformStrategy) {
+		panic("uniform strategy does not contain all score reqs")
+	}
+}
+
+// get the overall requirements from the policy and assign slots that'll fulfil them
+// TODO: this function should be changed in the future since it only supports stake reqs
+func CalcSlots(policy planstypes.Policy, minStake sdk.Int) []*PairingSlot {
+	// init slot array (should be as the number of providers to pair)
+	slots := make([]*PairingSlot, policy.MaxProvidersToPair)
+
+	// all slots should consider the stake, so we init them with stakeReq
+	stakeReq := StakeReq{MinStake: minStake}
+	stakeReqName := stakeReq.GetName()
+
+	for i := range slots {
+		reqMap := make(map[string]ScoreReq)
+		reqMap[stakeReqName] = stakeReq
+		slots[i] = NewPairingSlot()
+		slots[i].Reqs = reqMap
+	}
+
+	return slots
+}
+
+// group the slots
+func GroupSlots(slots []*PairingSlot) []*PairingSlot {
+	uniqueSlots := []*PairingSlot{}
+
+	if len(slots) == 0 {
+		panic("no pairing slots available")
+	}
+
+	uniqueSlots = append(uniqueSlots, slots[0])
```

Comment:
why do we need the special case for 0? we can start with an empty uniqueSlots array

Reply:
done. see https://github.com/lavanet/lava/commit/157c0df6080ba5ebb599d47c6873aca506b82b24